lxd/project/permissions.go

```diff
@@ -79,6 +85,27 @@ func AllowInstanceCreation(tx *db.ClusterTx, projectName string, req api.Instanc
 	return nil
 }
 
+// Check that we have not exceeded the maximum total alotted number of instances
+// for both containers and vms
+
+func checkTotalInstanceCountLimit(project *api.Project, totalInstanceCount int) error {
+	
+	overallValue, ok := project.Config["limits.instances"]
+	if ok {
+		limit, err := strconv.Atoi(overallValue)
+		if err != nil || limit < 0 {
+			return fmt.Errorf("Unexpected limits.instances value: %q", overallValue)
+		}
```

Comment:
Here we should just `return err` if it's `!= nil` and do the better validation in `validateTotalInstanceCountLimit` which should provide that message if strconv fails in it as well as ensure it's not `< 0`

Reply:
Here we were mimicking the general convention used for checking parameters as done in `checkInstanceCountLimit`. In our design, `validateTotalInstanceCountLimit` provides validation when the project parameters are changed (analogous to `validateInstanceCountLimit`) and `checkTotalInstanceCountLimit` providing validation when new instances are spun up (analogous to `checkInstanceCountLimit`). 
Is there a better approach that we could take? Should `validateTotalInstanceCountLimit` be called from `checkTotalInstanceCountLimit`?

Reply:
It may be a wrong pattern in the rest of the code that we'll want to sort out later.
Once a value is in the database, it's good, we don't need to validate it every time we read it back.

At the top, you've already defined a validation string as `validate.Optional(validate.IsUint32)`, so it's impossible for `limits.instances` to not be a valid uint32 coming out of the database.

`validateTotalInstanceCountLimit` is also called as part of that database record update, so that's where the `> 0` part of the validation should be.

`checkTotalInstanceCountLimit` can then move on knowing that `limits.instances` will always be a valid uint32 and will always be positive.